c10:intrusive_ptr, self assignment

DenisYaroshevskiy opened the pull request:
Summary:
In C++ books/sources, self assignment check is often considered a bad practice, since it is very very unlikely.

See, for example libc++ doesn't have it:
https://github.com/llvm/llvm-project/blob/cf94e0082e5e0a9be43a69d9fae588bc2aafab91/libcxx/include/__memory/shared_ptr.h#L651

How about we remove it?

Test Plan:
This check is like 1% of cycles assinged to intrusive_ptr::operator=
https://fburl.com/scuba/strobelight_services/9qqnrkdn

This is not a lot in purely cycles but since it's gpu machines, can be substantial

Differential Revision: D53471639

Skylion007 commented:
Source on it being bad practice to guard against this behavior? Clang-tidy actually recommends guarding against it for example: https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unhandled-self-assignment.html

DenisYaroshevskiy commented:
From clang 
> When true, the check will warn only if the container class of the copy assignment operator has any suspicious fields (pointer or C array). This option is set to true by default.

So it's not about just any self assignment

I read this suggestion originally in Exceptional C++ , can't find it online. 
Here are core guidelines: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#enforcement-101

DenisYaroshevskiy commented:
To be clear - the behaviour will stay the same with and without that check. Except the tradeoff is different: no branch in common case vs early exit in rare case.

ezyang commented:
@cyyever as the one who added the self assignment test, could you take a look?

ezyang commented:
@swolchok looking would be good too, I know at least some of the self assignment tests are in fact necessary because Scott added some of them.

DenisYaroshevskiy commented:
> some of the self assignment tests are in fact necessary 

I am not sure I understand this statement. As far as correctness is concerned, there are tests for this.

Scott replied on the internal diff. He was concerned if this would cause a performance regression in case people use extremely pathological workflows. That seems exceptionally unlikely to me but I have no hard data to support or reject this claim.

If you require that data to make this change, it is absolutely not worth it.
However I would suggest to avoid excessive caution. 

In general, if torch performance is a concern - the way to address it is to write benchmarks and optimize, not avoid changes.

ezyang commented:
I'm not too worried about perf, just whether or not it is correct.

> As far as correctness is concerned, there are tests for this.

This could be true, but it's not something I would take on faith since you can easily imagine the tests not testing self assignment, because they forgot to add. That being said, if swolchok thinks it is OK, I'm OK with it too.

DenisYaroshevskiy commented:
FYI: self assignment tests: https://github.com/pytorch/pytorch/blob/1c67f6cb269db5baaf2441cc39e4ebce0a813de3/c10/test/util/intrusive_ptr_test.cpp#L331-L346

cyyever commented:
I don't have specific opinions on self-assignment. Just make the linter happy if the check is removed.

DenisYaroshevskiy commented:
> PS: C++ Core Guidelines has two rules
[C.62: Make copy assignment safe for self-assignment](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-copy-self)
and
[C.65: Make move assignment safe for self-assignment](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-move-self)
Therefore may be the new trend in the community is to check self-assignment in user-defined operator== ?

Make safe means it should be OK - which it is. But copy & swap does that.
Here is from CppCoreGuidelines.
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#enforcement-101

cyyever commented:
@ezyang I am fine with the changes. And if the code is broken somewhere after merging, that means some hidden logical bugs cause instrusive_ptr self-assignment and we should find them out.

DenisYaroshevskiy commented:
@pytorchbot merge

DenisYaroshevskiy commented:
@pytorchbot merge

DenisYaroshevskiy commented:
@pytorchbot merge

DenisYaroshevskiy commented:
@pytorchbot merge

DenisYaroshevskiy commented:
@pytorchbot label "topic: not user facing"

DenisYaroshevskiy commented:
Can someone approve the workglows for me?

cyyever commented on c10/util/intrusive_ptr.h:
```diff
@@ -392,10 +392,9 @@ class intrusive_ptr final {
     return *this;
   }
 
+  // Assignment is implemented using copy and swap. That's safe for self assignment.
```
To pass the linter, these should be changed to
```
// Assignment is implemented using copy and swap. That's safe for self
// assignment.
// NOLINTNEXTLINE(bugprone-unhandled-self-assignment)
```

DenisYaroshevskiy committed:
c10:intrusive_ptr, self assignment (#119275)

Summary:

In C++ books/sources, self assignment check is often considered a bad practice, since it is very very unlikely.

See, for example libc++ doesn't have it:
https://github.com/llvm/llvm-project/blob/cf94e0082e5e0a9be43a69d9fae588bc2aafab91/libcxx/include/__memory/shared_ptr.h#L651

How about we remove it?

Test Plan:
This check is like 1% of cycles assinged to intrusive_ptr::operator=
https://fburl.com/scuba/strobelight_services/9qqnrkdn

This is not a lot in purely cycles but since it's gpu machines, can be substantial

Differential Revision: D53471639

DenisYaroshevskiy commented:
```
>>> Lint for test/onnx/test_fx_op_consistency.py:

  Warning (UFMT) format
    Run `lintrunner -a` to apply this patch.

    You can run `lintrunner -a` to apply this patch.

    1665  1665 | def _torch_size_flatten_spec(d: List[Any], spec: Any) -> List[Any]:
    1666  1666 |     return [d[i] for i in range(spec.num_children)]
    1667  1667 | 
          1668 |+
    1668  1669 | torch.fx._pytree.register_pytree_flatten_spec(
    1669  1670 |     torch.Size,
    1670  1671 |     _torch_size_flatten_spec,

+ echo ''
```

This is not the code I touched.

cyyever commented:
@pytorchmergebot -i

cyyever commented:
@pytorchmergebot merge -i
